### PR TITLE
fetching raghava trello data to gold table

### PR DESCRIPTION
**Pull request by nelabhotlaR**
python raghava_nelbo_trello_to_gold.py --start_date='2022-08-01' --end_date='2023-02-20'

loading settings :: url = jar:file:/home/lakeuser/lakehouse_project/venv/lib/python3.8/site-packages/pyspark/jars/ivy-2.5.0.jar!/org/apache/ivy/core/settings/ivysettings.xml
Ivy Default Cache set to: /home/lakeuser/.ivy2/cache
The jars for the packages stored in: /home/lakeuser/.ivy2/jars
io.delta#delta-core_2.12 added as a dependency
:: resolving dependencies :: org.apache.spark#spark-submit-parent-d50c196c-7782-4293-b6ec-3c81ffd334e4;1.0
	confs: [default]
	found io.delta#delta-core_2.12;2.1.1 in central
	found io.delta#delta-storage;2.1.1 in central
	found org.antlr#antlr4-runtime;4.8 in central
	found org.codehaus.jackson#jackson-core-asl;1.9.13 in central
:: resolution report :: resolve 214ms :: artifacts dl 8ms
	:: modules in use:
	io.delta#delta-core_2.12;2.1.1 from central in [default]
	io.delta#delta-storage;2.1.1 from central in [default]
	org.antlr#antlr4-runtime;4.8 from central in [default]
	org.codehaus.jackson#jackson-core-asl;1.9.13 from central in [default]
	---------------------------------------------------------------------
	|                  |            modules            ||   artifacts   |
	|       conf       | number| search|dwnlded|evicted|| number|dwnlded|
	---------------------------------------------------------------------
	|      default     |   4   |   0   |   0   |   0   ||   4   |   0   |
	---------------------------------------------------------------------
:: retrieving :: org.apache.spark#spark-submit-parent-d50c196c-7782-4293-b6ec-3c81ffd334e4
	confs: [default]
	0 artifacts copied, 4 already retrieved (0kB/5ms)
23/02/20 17:08:17 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).
2023-02-20 17:08:29.534 | INFO     | __main__:get_activity_cards:36 - Fetched the refined cards data from silver table
2023-02-20 17:08:37.000 | SUCCESS  | __main__:get_activity_cards:52 - 
 Saved raghava cards data to /home/lakeuser/lakehouse_project/data_tables/raghava_cards_gold_path
2023-02-20 17:08:37.000 | INFO     | __main__:get_activity_cards:54 - Saved raghava cards data for boardto /home/lakeuser/lakehouse_project/data_tables/raghava_cards_gold_path table

### CODE REVIEW

File: jobs/raghava_nelbo_trello_to_gold.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
+"""
+This script will fetch all the cards of a trello board that according to the below criteria:
```
Comment (sravantit25) [inline jobs/raghava_nelbo_trello_to_gold.py]: This description needs to be changed. It is of the noactivity cards script. And, what this script does is different.

File: jobs/raghava_nelbo_trello_to_gold.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+"""
+
+import os
+import sys
+from datetime import datetime, timedelta
+import argparse
+from loguru import logger
+from pyspark.sql.functions import col
+sys.path.append(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))
+from helpers import spark_helper as sh, common_functions as cf, trello_functions as tf
```
Comment (sravantit25) [inline jobs/raghava_nelbo_trello_to_gold.py]: Can you remove common_functions and trello_functions. They are not needed.

File: jobs/raghava_nelbo_trello_to_gold.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+from config import delta_table_conf as dc
+
+def get_activity_cards(start_date,end_date):
```
Comment (sravantit25) [inline jobs/raghava_nelbo_trello_to_gold.py]: Function name needs to change. This is not related to activity cards. Can instead say member_cards

File: jobs/raghava_nelbo_trello_to_gold.py [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+    """
+    Fetch all the cards data between provided date range and
+    active from past specified days and save it into a Delta Lake Gold table
+    :param spark: Spark session
+    :param board_id: trello board id
+    :param board_name: trello board name
+    :param no_of_days: number of days the cards are not active
+    :return none
```
Comment (sravantit25) [inline jobs/raghava_nelbo_trello_to_gold.py:23]: The description needs to be changed. It is that of noactivity

File: jobs/raghava_nelbo_trello_to_gold.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    """
+    path = dc.raghava_cards_gold_path
+    try:
+        spark, context = sh.start_spark()
+
+        # Fetch the cards data from the refined cards silver delta table
+        refined_cards_data = spark.read.format(
+            "delta").load(dc.cards_silver_table_path)
```
Comment (sravantit25) [inline jobs/raghava_nelbo_trello_to_gold.py]: The Silver cards table used here is incorrect. It should be unique_silver_table. Instructions present in the Google doc.

File: jobs/raghava_nelbo_trello_to_gold.py [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+        logger.info(f'Fetched the refined cards data from silver table')
+
+        # trim date using datetime
+        start_date = datetime.strptime(start_date,'%Y-%m-%d')
+        end_date = datetime.strptime(end_date,'%Y-%m-%d')
+        # filter the data based on name, start date and end date
+        filtered_data = refined_cards_data.filter(col("card_members").contains('raghavanelabhotla') & (col("dateLastActivity") >= start_date) & (col("dateLastActivity") <= end_date))
```
Comment (sravantit25) [inline jobs/raghava_nelbo_trello_to_gold.py]: This script errored out for me. When I checked, I saw that your name is present as raghava.nelabhotla. But here you are using raghavanelabhotla. Please change it to something like %raghava% so that it will work in any case.

File: jobs/raghava_nelbo_trello_to_gold.py [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+        #Select the required columns
+        active_cards_data = filtered_data.select(
```
Comment (sravantit25) [inline jobs/raghava_nelbo_trello_to_gold.py]: Can you rename this to member_cards_data.

**Review by sravantit25 — CHANGES_REQUESTED** [review]

File: jobs/raghava_nelbo_trello_to_gold.py (hunk 1 above)
Comment (nelabhotlaR) [reply jobs/raghava_nelbo_trello_to_gold.py]: """
This script will fetch all the cards of a Trello board according to the defined criteria:
This script fetches all the Trello cards that I have worked on over a period of time 
and place it in a Gold delta table
This script is scheduled to run twice a week
"""

File: jobs/raghava_nelbo_trello_to_gold.py (hunk 2 above)
Comment (nelabhotlaR) [reply jobs/raghava_nelbo_trello_to_gold.py]: Removed unwanted imports

File: jobs/raghava_nelbo_trello_to_gold.py (hunk 3 above)
Comment (nelabhotlaR) [reply jobs/raghava_nelbo_trello_to_gold.py]: modified the function definition to get_memeber_cards

File: jobs/raghava_nelbo_trello_to_gold.py (hunk 4 above)
Comment (nelabhotlaR) [reply jobs/raghava_nelbo_trello_to_gold.py:23]: removed unwanted lines

File: jobs/raghava_nelbo_trello_to_gold.py (hunk 5 above)
Comment (nelabhotlaR) [reply jobs/raghava_nelbo_trello_to_gold.py]: changed the table to cards_unique_silver_table_path

File: jobs/raghava_nelbo_trello_to_gold.py (hunk 6 above)
Comment (nelabhotlaR) [reply jobs/raghava_nelbo_trello_to_gold.py]: as suggested changed to "%raghavanelabhotla%"

File: jobs/raghava_nelbo_trello_to_gold.py (hunk 7 above)
Comment (nelabhotlaR) [reply jobs/raghava_nelbo_trello_to_gold.py]: changed the variable name to member_cards_data

File: jobs/raghava_nelbo_trello_to_gold.py (hunk 6 above)
Comment (sravantit25) [reply jobs/raghava_nelbo_trello_to_gold.py]: @nelabhotlaR I think &raghava....% will work with like.
If you are using contains, then you should give only the name. So, the line needs to change to either use like(%raghava%) or contains(raghava.nelabhotla).
Because right now the script has not fetched any data.

I changed the line to below: and then it fetched some data.

filtered_data = refined_cards_data.filter((col("card_members").contains("raghava.nelabhotla")) & (col("dateLastActivity") >= start_date) & (col("dateLastActivity") <= end_date))

**Review by sravantit25 — CHANGES_REQUESTED** [review]
@nelabhotlaR Can you please check the line where you are making the condition. I have replied to the particular comment.

Need to use either contains("raghava.nelabhotla") or like(%raghava%)

File: jobs/raghava_nelbo_trello_to_gold.py (hunk 6 above)
Comment (nelabhotlaR) [reply jobs/raghava_nelbo_trello_to_gold.py]: Modified the step to use contains(raghava.nelabhotla)

Comment (sravantit25) [reply jobs/raghava_nelbo_trello_to_gold.py]: thanks @nelabhotlaR Tested and works fine now. Merging the PR.
